From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Wed, 30 Jul 2014 14:07:30 +0200
Subject: [PATCH] gpu: ipu-v3: Add CPMEM hacks for V4L2 output drivers

The V4L2 output drivers need some information about running IDMAC channels.
Provide it to them.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 drivers/gpu/ipu-v3/ipu-cpmem.c | 39 +++++++++++++++++++++++++++++++++++++++
 include/video/imx-ipu-v3.h     |  4 ++++
 2 files changed, 43 insertions(+)

diff --git a/drivers/gpu/ipu-v3/ipu-cpmem.c b/drivers/gpu/ipu-v3/ipu-cpmem.c
index 50b9cef17110..e2efea5b7e73 100644
--- a/drivers/gpu/ipu-v3/ipu-cpmem.c
+++ b/drivers/gpu/ipu-v3/ipu-cpmem.c
@@ -885,6 +885,45 @@ void ipu_cpmem_dump(struct ipuv3_channel *ch)
 }
 EXPORT_SYMBOL_GPL(ipu_cpmem_dump);
 
+/*
+ * The following three functions are a major hack for the V4L2 overlay driver
+ */
+void ipu_cpmem_get_base_resolution(struct ipu_soc *ipu, u32 *width, u32 *height)
+{
+	struct ipuv3_channel *ch = &ipu->channel[IPUV3_CHANNEL_MEM_BG_SYNC];
+
+	*width = ipu_ch_param_read_field(ch, IPU_FIELD_FW) + 1;
+	*height = ipu_ch_param_read_field(ch, IPU_FIELD_FH) + 1;
+}
+
+int ipu_cpmem_save(struct ipuv3_channel *ch, void **cpmem, u32 *eba)
+{
+	struct ipu_ch_param __iomem *p = ipu_get_cpmem(ch);
+
+	if (*cpmem)
+		kfree(cpmem);
+	*cpmem = kzalloc(sizeof(**cpmem), GFP_KERNEL);
+	if (!*cpmem)
+		return -ENOMEM;
+
+	memcpy(*cpmem, p, sizeof(struct ipu_ch_param));
+	*eba = ipu_ch_param_read_field(ch, IPU_FIELD_EBA0) << 3;
+
+	return 0;
+}
+
+void ipu_cpmem_restore(struct ipuv3_channel *ch, void **cpmem)
+{
+	struct ipu_ch_param __iomem *p = ipu_get_cpmem(ch);
+
+	if (!*cpmem)
+		return;
+
+	memcpy(p, *cpmem, sizeof(struct ipu_ch_param));
+	kfree(*cpmem);
+	*cpmem = NULL;
+}
+
 int ipu_cpmem_init(struct ipu_soc *ipu, struct device *dev, unsigned long base)
 {
 	struct ipu_cpmem *cpmem;
diff --git a/include/video/imx-ipu-v3.h b/include/video/imx-ipu-v3.h
index e6f22a5489b0..1a6efe3de90a 100644
--- a/include/video/imx-ipu-v3.h
+++ b/include/video/imx-ipu-v3.h
@@ -273,6 +273,10 @@ void ipu_cpmem_set_yuv_planar(struct ipuv3_channel *ch,
 int ipu_cpmem_set_fmt(struct ipuv3_channel *ch, u32 drm_fourcc);
 int ipu_cpmem_set_image(struct ipuv3_channel *ch, struct ipu_image *image);
 void ipu_cpmem_dump(struct ipuv3_channel *ch);
+/* Hacks for V4L2 overlay driver */
+void ipu_cpmem_get_base_resolution(struct ipu_soc *ipu, u32 *width, u32 *height);
+int ipu_cpmem_save(struct ipuv3_channel *ch, void **cpmem, u32 *eba);
+void ipu_cpmem_restore(struct ipuv3_channel *ch, void **cpmem);
 
 /*
  * IPU Display Controller (dc) functions
